src/common/serialize.rs

```diff
@@ -14,6 +14,20 @@ pub trait BinarySerializable: fmt::Debug + Sized {
     fn deserialize<R: Read>(reader: &mut R) -> io::Result<Self>;
 }
 
+pub trait DeserializeFrom<T: BinarySerializable> {
+    fn deserialize(&mut self) -> io::Result<T>;
+}
+
+/// Implement deserialize from &[u8] for all types which implement BinarySerializable.
+///
+/// TryFrom would actually be preferrable, but not possible because of the orphan
```

Comment:
I am not sure this trait makes the code any simpler.

If try from what the target, why not take mut self?

Reply:
It's used below, imo quite nice to use
```     
   let (footer_len, footer_magic_byte): (u32, u32) = file
            .slice_from_end(footer_metadata_len)
            .read_bytes()?
            .as_ref()
            .deserialize()?;
```

It's `&mut self` because the trait wants `&mut  Read`

with try_from it would be converting from `&[u8]` to `T` I think.